USWDS - Input & Prefix/Suffix: Add disabled styles. Resolves #4913

mahoneycm opened the pull request:
## Description

[Preview →](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/cm-prefix-suffix-disable/?path=/story/components-form-inputs-input-prefix-or-suffix--prefix)

Resolves #4913 

### Input disabled styling

Currently, disabled form elements other than `button` rely on default browser styles to give visual queues that form elements are disabled. Chrome and Edge very lightly dim the field, but is sometimes hard to notice. Other browsers such as Safari and Firefox do not add any visual indication that these fields are inaccessible to the user. 

@amyleadem summed it up very nicely in her comment on #4322 which addresses this issue for `select / dropdown` components

> @thisisdano @mejiaj I have included a screenshot that showcases disabled visual cues in Chrome, Firefox, Safari, and Edge for OS. This screenshot displays the following:
> 
> * Standard .usa-select element
> * Disabled .usa-select element
> * Disabled select element without .usa-select class
> * Standard .usa-input element
> * Disabled .usa-input element
> * Disabled input element without .usa-input class
> 
> ![Image](https://user-images.githubusercontent.com/93996430/193339979-915af696-7857-4e10-84a3-60004b40b606.png)
> 
> I also have a link to a [live demo](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-test-disabled-select/iframe.html?args=&id=components-form-inputs-select--select&viewMode=story) in PR #4969 in case it is hard to glean the style differences from the screenshot.
> 
> Things to note:
> 
> * Each browser has its own set of default styles for disabled form elements.
> * USWDS overrides some default disabled styles in a way that makes it less clear visually that the element is disabled.
> 
> I lean towards wanting to take a look at developing `disabled` styles for USWDS form elements that provide clearer visual cues that the element is disabled. However, we would need to be careful of poor contrast ratio and look into accessibility implications.

 I agree with Amy in leaning towards developing our own `disabled` styles to have a consistent and easy-to-understand style across all USWDS form elements and browsers.

This PR adds `disabled` styles for `input` elements but can be extended to other form elements such as text fields, drop downs, and combo boxes if we chose to take this route.

### Prefix / Suffix

As mentioned in #4913, input prefix/suffix does not inherit the input disabled style. This can be resolved by creating a `usa-input-group--disabled` style. The usage would be consistent with how we implement the error state with `usa-input-group--error`. 

Prefix / Suffix are the only components that make use of the `usa-input-group` class so there is no worry about the styles breaking another component.

## Screenshots

**Before**

Prefix with disabled input:
<img width="251" alt="image" src="https://user-images.githubusercontent.com/25211650/197559691-9b12960a-a4cc-4413-b95a-9d9c5238dd8c.png">

**After**
Prefix with `usa-input-group--disabled` and disabled input:

<img width="251" alt="Screen Shot 2022-11-28 at 9 43 57 AM" src="https://user-images.githubusercontent.com/25211650/204306092-c5995654-92a2-4817-8f7e-b2c67270aaf0.png">

mejiaj reviewed (changes requested):
No issues found with input prefix / suffix specifically. Added comments based on additional findings.

**Tested**
- [x] Code style and formatting
- [x] Input prefix/suffix with disabled attribute
- [x] Adding `disabled` attribute to all inputs found in components under `Form Inputs`

mahoneycm commented on packages/usa-input/src/styles/_usa-input.scss:
```diff
@@ -3,6 +3,10 @@
   @extend %block-input-styles;
+
+  &:disabled {
+    background-color: color("disabled-light");
```
I am on team "add disabled styles" to our other form inputs. I wanted to PR this out before getting too ahead of myself and adding them to all the form elements. In my PR I wrote

> This PR adds disabled styles for input elements but can be extended to other form elements such as text fields, drop downs, and combo boxes if we chose to take this route.

If you want me to start playing around with adding them let me know and I'll get cookin'!

mahoneycm commented on packages/usa-input-prefix-suffix/src/usa-input-prefix.twig:
```diff
@@ -19,4 +19,14 @@
     <input type="text" id="example-input-prefix-error" class="usa-input" pattern="[0-9]*" inputmode="numeric">
   </div>
+
+  <label class="usa-label" for="example-input-prefix-error">Credit card number (Disabled)</label>
```
Resolved in [83f5d39](https://github.com/uswds/uswds/pull/5004/commits/83f5d391496b02d790e918ddd32f41bf463e61b1)

mahoneycm commented on packages/usa-input-prefix-suffix/src/usa-input-prefix.twig:
```diff
@@ -19,4 +19,14 @@
+        <use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="./img/sprite.svg#credit_card"></use>
+      </svg>
+    </div>
+    <input type="text" id="example-input-prefix-error" class="usa-input" pattern="[0-9]*" inputmode="numeric" disabled>
```
Resolved in [83f5d39](https://github.com/uswds/uswds/pull/5004/commits/83f5d391496b02d790e918ddd32f41bf463e61b1)

mahoneycm commented on packages/usa-input-prefix-suffix/src/styles/_usa-input-prefix-suffix.scss:
```diff
@@ -13,6 +13,10 @@
   }
 
+  &--disabled {
+    background-color: color(disabled-light);
```
Resolved in [123d806](https://github.com/uswds/uswds/pull/5004/commits/123d8061a7ee20cf9afcb7729e1a9f266af3d2f6)

mahoneycm commented:
## Added styles for text input
_Screenshots go from Active inputs, default styles when disabled, new styles with USWDS disabled_
![image](https://user-images.githubusercontent.com/25211650/199309668-3e645efc-020c-4a49-965d-488689f28fdd.png)

This can easily be added to other inputs such as `select` or `combo-box`.

What do you all think @mejiaj @thisisdano @amyleadem ? Something to consider implementing?

mejiaj reviewed (approved):
Looks good on this component. Let's create a new issue (if there isn't one already) for extending this to other form inputs.

mahoneycm commented:
Issue created here: https://github.com/uswds/uswds/issues/5051

amyleadem commented:
@mahoneycm 
Looks good. I am a bit torn on the styling because it doesn't read as disabled to me from a visual perspective. But I understand that we likely need to maintain accessible color contrast ratios and I like that it matches the disabled date picker styles.

I added a comment to https://github.com/uswds/uswds/issues/5051 about possibly standardizing USWDS disabled styles since there are differing levels of contrast in disabled styles across the design system. 

Additionally, our radio and checkbox disabled styles have added `cursor: not-allowed` in their styles. Would it make sense to add that here as well?

mahoneycm commented:
Originally I had set the color to `disabled-light` because it was closer to Chromes disabled styles.  I went ahead and switched it to `disabled` so it was consistent to other disabled elements and added `cursor: not-allowed` as recommended!

amyleadem reviewed (changes requested):
Hi @mahoneycm,
I remembered that we should add the same styles for `aria-disabled` as well (like in PR https://github.com/uswds/uswds/pull/4783). 

I also noticed that date picker no longer has styles for `aria-disabled`. Can you take a look to see if your fix will repair that?

Thanks!

mahoneycm commented:
@amyleadem  I noticed that other forms include `aria-disabled` styles while working on my other disabled forms PR and I planned on adding them in, so I'll do that for this PR as well!

As for date picker, I'll look into it as part of #5051

mahoneycm commented:
@amyleadem  I added `aria-disabled` styles and an example within the component previews. 

`date-picker` still has `aria-disabled` on this branch, but I'll make sure it's included on #5051

mahoneycm commented:
@thisisdano Ready for final review!

amyleadem reviewed (changes requested):
:warning: This fix conflicts with the effort to remove disabled classes in https://github.com/uswds/uswds/pull/5128. Is there an alternate method for styling this component?

amyleadem commented:
Going to change the status on this issue to "Review" to allow for evaluation.

amyleadem commented on packages/uswds-core/src/styles/mixins/utilities/_disabled.scss:
```diff
@@ -0,0 +1,6 @@
+@use "../../functions/utilities/color" as *;
+
```
Can you add a comment with a brief description of what this mixin does?

mahoneycm replied:
Resolved in https://github.com/uswds/uswds/pull/5004/commits/7963934845140e22cc671938cba057fcd5eab31d

amyleadem replied:
Sorry - I meant it would be helpful to have description for what `@mixin u-disabled` does. I'll put my comment in a better place next time!

mahoneycm replied:
GOSH hindsight is 20/20 I feel like a goober for not getting that the first time 😅

Resolved in https://github.com/uswds/uswds/pull/5004/commits/25f19505f450e6a797431c50c684ebffc5979ea0 !

amyleadem reviewed (changes requested):
I like this styling solution and I like the clarity of the variables up top, @mahoneycm! I added a few comments for your review. 

Additionally, what do you think about the possibility of removing the `index.js` file now that the input is full width and absolutely positioned? I might be missing something, but with a little tweaking it feels like we can handle focus with CSS exclusively now. Curious to hear your thoughts!

mahoneycm commented on packages/usa-input-prefix-suffix/src/styles/_usa-input-prefix-suffix.scss:
```diff
@@ -1,11 +1,16 @@
+// Padding helpers
+$ICON_WIDTH: 3;
+$PREFIX_SUFFIX_PADDING: 1;
+$PREFIX_SUFFIX_OFFSET: $ICON_WIDTH + ($PREFIX_SUFFIX_PADDING * 2) + 1; // 1 is padding from input sytles
```
Resolved in https://github.com/uswds/uswds/pull/5004/commits/8d817939ab88f37b870af21639d7a98bcf851760

mahoneycm commented on packages/usa-input-prefix-suffix/src/styles/_usa-input-prefix-suffix.scss:
```diff
@@ -1,11 +1,16 @@
+// Padding helpers
+$ICON_WIDTH: 3;
+$PREFIX_SUFFIX_PADDING: 1;
+$PREFIX_SUFFIX_OFFSET: $ICON_WIDTH + ($PREFIX_SUFFIX_PADDING * 2) + 1; // 1 is padding from input sytles
```
I agree! I liked it better without the `+ 1` as well. I removed it and think it looks much better and leaves `prefix_suffix_offset` a bit cleaner

mahoneycm replied:
Resolved in https://github.com/uswds/uswds/pull/5004/commits/4b020e39b6779ceea9cd6e640dd53eb06390269e

mahoneycm commented on packages/usa-input-prefix-suffix/src/usa-input-prefix.twig:
```diff
@@ -19,4 +19,24 @@
+  </div>
+
+  <label class="usa-label" for="example-input-prefix-aria-disabled">Credit card number (aria-disabled)</label>
+  <div class="usa-input-group usa-input-group--disabled">
```
Resolved in https://github.com/uswds/uswds/pull/5004/commits/5581cdd15235c300eb98506450c5f7fdf92243ea

mahoneycm commented:
@amyleadem Thanks for the great feedback! Resolved all of your comments if you'd like to give it a double check!

mahoneycm commented on packages/usa-input-prefix-suffix/src/styles/_usa-input-prefix-suffix.scss:
```diff
@@ -21,28 +24,52 @@
     @include u-border($theme-input-state-border-width, "success");
   }
 
+  &--error,
+  &--success {
+    & input:focus {
+      outline-offset: units($theme-input-state-border-width);
+    }
+  }
```
Currently, the focus outline is placed outside of the succes/error border. This allows for the same styling

![image](https://user-images.githubusercontent.com/25211650/218498843-936632e3-a1d5-4f26-80d2-04f9203da40f.png)

If we prefer to have the focus cover the success/error border, we can remove these lines

amyleadem replied:
I think the the outline-offset is a good consideration. Nice call.

amyleadem replied:
I did notice that the input border (1px) is now covered by the focus outline, whereas it was visible before. It is a small thing, but just wanted to flag the visual difference. Not sure if it needs any action.

amyleadem commented on packages/usa-input-prefix-suffix/src/styles/_usa-input-prefix-suffix.scss:
```diff
@@ -21,28 +24,52 @@
 
+.usa-input-prefix + input {
+  @include u-padding-left($prefix_suffix_offset);
+  @include u-padding-right(1);
```
Could this be defined with `$prefix_suffix_padding`?
```suggestion
  @include u-padding-right($prefix_suffix_padding);
```

mahoneycm replied:
Resolved in https://github.com/uswds/uswds/pull/5004/commits/5db72ab757a65629f44993a4b6b99d08e58f0b11

amyleadem reviewed (changes requested):
Looking good! I had a couple of comments for you &mdash; just some little things.

mejiaj reviewed (changes requested):
@mahoneycm some minor comments, otherwise LGTM.

mahoneycm commented:
@mejiaj Thanks for sharing! Always happy to have a method to the madness and organize our files.

I did some cleanup and pushed it up if you'd like to re review! Let me know if there's anything I can organize further!

mejiaj reviewed (approved):
Looks good! Thank you for your patience on this. Tested in Firefox, Chrome, and Safari.

Also, I went ahead and ran prettier sass task.

---

Odd, the [original test](https://app.circleci.com/pipelines/github/uswds/uswds/4665/workflows/43a9b697-8778-42bd-b22e-3b3878d8858d) failed in CircleCI due to a timeout. Wasn't able to reproduce locally and resolved on re-run.

mahoneycm commented:
Work implemented in #5063
